### ISSUE TITLE
SpirvToolsDisassemble does not allow choosing alternative `spv_target_env`.

### ISSUE THREAD

**Issue by phniix** [KhronosGroup/glslang#2405]
#### ;tldr

The utility function, `glslang::SpirvToolsDisassemble`, does not allow choosing an alternative SPIR-V context. We need to expand this.

Description
--- 
The current interface does not provide the flexibility to allow the evaluation to be performed in other SPIR-V environments. It would be good if a new interface was provided, to allow for this.

Additionally, the current API's fixed choice, `SPV_ENV_UNIVERSAL_1_3` is not documented, and requires looking into the source code to determine this fact. It would be good if the documentation was improved at the API level, to at least not necessitate digging into the source code to know this.

Proposal
---
I propose a new API interface, as opposed to leveraging the existing interface. I feel that such a distinction is worthwhile.

The new API interface would be akin to the existing one, with an additional parameter.
```
void SpirvToolsDisassemble(std::ostream& out, const std::vector<unsigned int>& spirv,
           spv_target_env requested_context);
```

I'm not overly stressed should a preference lean more heavily towards only on a single interface:
```
void SpirvToolsDisassemble(std::ostream& out, const std::vector<unsigned int>& spirv,
           spv_target_env requested_context=SPV_ENV_UNIVERSAL_1_3);
```

The existing API will also get a brand new coat of "documentation paint" for the `SPV_ENV_UNIVERSAL_1_3` usage.

Reference Implementation
---
A reference implementation for the above proposal exists here: https://github.com/phniix/glslang/tree/SpirvToolsDisassemble-Improvements

**Comment by phniix** [comment]
Updates made based on feedback from @johnkslang. Thanks!!
New PR #2408 tidies up the reference implementation.

**Comment by phniix** [comment]
@johnkslang Should this get opened up again, because of the situation outlined in https://github.com/KhronosGroup/glslang/issues/2413

I asked a bunch of questions to yourself, and @timvpGoogle there upon discovery of that issue.
There is here: https://github.com/KhronosGroup/glslang/issues/2413#issuecomment-706869986

Following all those questions, and whether this opens up again or we spin off a new one...

What's the best way to proceed ?

**Comment by johnkslang** [comment]
I reopened #2413 for further discussing this.

**Comment by phniix** [comment]
Issue https://github.com/KhronosGroup/glslang/issues/2434 addressed what caused this issue to be reverted a couple of times.

(@johnkslang) I'm wondering if this issue should now be re-opened, or a new issue created, which references this one. With https://github.com/KhronosGroup/glslang/pull/2429 we now have some confidence and security for the presubmit/CI, to give us feedback on how the code change and its build script modifcations, trickle down the entire dependency chain of presubmits.

I'm not 100% on the fix for GN, but my guess described here, https://github.com/KhronosGroup/glslang/issues/2413#issuecomment-711035068 , feels like the right approach. 

(@timvpgoogle) Do you think someone in your team with GN experience may perhaps look at my proposed solution, and is able to contribute some pointers to help us fastrack getting this feature out to the community ?

**Comment by timvpGoogle** [comment]
@ben-clayton May be able to help take a look.

**Comment by phniix** [comment]
Thanks for suggesting someone on your team ! That will really help us all out. (@timvpGoogle)

@ben-clayton Would you be able to take a look at the comments above, and perhaps provide assistance with your GN expertise ?

**Comment by ben-clayton** [comment]
Hi @phniix,

I'm afraid I'm no GN expert, and it's not particularly clear to me what exactly needs doing here (there's a lot of information scattered across several threads). You mentioned a feature you'd like fast tracked. Please can you summarize exactly what you need?

With that said, I'm not sure if I'm the best person to tackle this - I've chipped in to add kokoro presubmits that were missing - but there's now debate around whether kokoro is the right presubmit / CI solution for glslang. It is possible we may wish to migrate to another system, in which case we should probably hold off until this decision is made. 

Cheers,
Ben

**Comment by phniix** [comment]
Hi Ben,

Firstly, I'd like to make sure that you have nothing to fear at all! No dramas, and truly appreciative of your questions and follow-up interest :)

Yes, it is starting to get a little bit spaghetti-like with the threads, as my thought was to prevent the repetition, but as you say, and as I now recognise, that's not the easiest to follow. Thank you for pointing that out :)

In terms of what feature, it is specifically this one here described in this issue. We have PRs that have been accepted, and then reverted, due to GN build dramas. We've had this in the can, ready to go, and keep hitting stumbling blocks.

I think your work for the kokoro presubmit is great, and will help us get this moving, as the GN presubmit should signal whether or not we can build glslang with the changes in this PR.

Now, not being versed in GN, and only taking a fairly decent guess at what needs to be done, I think (refer to https://github.com/KhronosGroup/glslang/issues/2413#issuecomment-711035068 above), we need to:

- Add: `include_dirs += [ "${spirv_tools_dir}/include" ]` to `if (invoker.enable_opt) {` in `glslang/BUILD.gn`
- Add: `include_dirs += [ "${spirv_tools_dir}/include" ]` to `executable("glslang_validator") {` in `glslang/BUILD.gn`
- Add: `include_dirs += [ "${spirv_tools_dir}/include" ]` to `executable("spirv-remap") {` in `glslang/BUILD.gn`

I'm definitely not following the idiomatic, easier to ask for forgiveness than permission principle here, and more closely aligned to the look before you leapin principle, in an effort to avert any dramas in the GN toolchain.

I could open up another PR that builds off of https://github.com/KhronosGroup/glslang/pull/2408 and then add the above lines in `glslang/BUILD.gn`, but not 100% sure that is what needs to get done. 

If I was to do the above, or perhaps someone with a GN build environment at their fingertips, would this correctly gives us a pass/fail result through the GN presubmit ?

As an aside, I think the discussion of what the right approach is for GN presubmit is orthogonal to getting the  feature described in this github issue, out to the community.

Does that help clarify ?

I look forward to hearing from you, or any other members of the GN/chromium team that can help fortify our forward progress.

All the best,
pheonix

(@ben-clayton)

**Comment by phniix** [comment]
@johnkslang, thanks a bunch for morphing this issue into a question which builds upon the learnings we had from the original. It's been a journey to date, and we are definitely making great progress on the fringe issues that we have unknowingly uncovered through this issue, that have a fundamental impact on this library's progress. This is great stuff. Thanks mate.

**Comment by arcady-lunarg** [comment]
Thanks to Google, we have CI coverage of GN builds via kokoro, so I think this issue can finally be put to rest.
